Improve Translator caching

rpkamp opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #27600
| License       | MIT
| Doc PR        | N/A

Add DirectoryResources to MessageCatalogues when loaded.

So that when a file is added to one of the directories the cache for all MessageCatalogues will be invalidated.

All directories must be added to all MessageCatalogues because we can't predict for which locale files will be added to each individual directory.

Also, now that the translator keeps track of its own directories for caching the container no longer needs to it. This means that when a translation changes or is added the container no longer needs to be fully rebuilt, saving a considerable amount of time (compilation time went down from ~4 seconds to ~1 second on each translation change/add in our project).

rpkamp committed:
Improve Translator caching

rpkamp commented:
Rebased and force pushed to fix merge conflict.

Could I get a review here please? :slightly_smiling_face:

rpkamp commented:
Rebased again.

Could you have a look please @nicolas-grekas? :slightly_smiling_face:

nicolas-grekas reviewed (commented):
Looks good thanks. Here are some comments. Would it be possible to add some test cases also?

rpkamp commented on src/Symfony/Bundle/FrameworkBundle/Translation/Translator.php:
```diff
@@ -120,6 +129,16 @@ protected function initializeCatalogue($locale)
         parent::initializeCatalogue($locale);
     }
 
+    protected function doLoadCatalogue($locale): void
```
No that won't work, because you'd also need this code triggered when `initializeCacheCatalogue` is called instead of `initializeCatalogue` (i.e., when a `cache_directory` is set).

rpkamp commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -1085,7 +1094,10 @@ private function registerTranslatorConfiguration(array $config, ContainerBuilder
-                ['resource_files' => $files]
+                [
+                    'resource_files' => $files,
+                    'scanned_directories' => \array_merge($dirs, $nonExistingDirs),
```
Which `$dir` do you mean?

sstok replied:
I guess `$defaultDir`? But this is already tracked.

rpkamp commented:
@nicolas-grekas I've rebased onto master and stopped using `ContainerBuilder::fileExists()` for translation resources altogether since the container no longer needs to track any changes in translation files - catalogues are fully handling that now.

This prevents a full container rebuild if any of the translation directories it was tracking but did not exist is created - it will only rebuild the Catalogue in that case, which is much faster.

I've also added some tests. 

Could you check again please? :slightly_smiling_face:

rpkamp commented:
Failing tests are unrelated (I think).

welcoMattic commented:
Hi @rpkamp, I have a similar issue that your PR could solve.  [Here](https://github.com/php-translation/symfony-bundle/issues/198) the use case and the steps to reproduce it. I'll try your PR with my scenario to test if it solves the problem.

Do you have a scenario to be reproduced? 

Thanks

rpkamp commented:
This PR makes sure that catalogues also track directories that were found empty when scanning for translation files, where they didn't before.

So yes, if I understand correctly this PR is also a fix for your problem, @welcoMattic :slightly_smiling_face:

rpkamp commented:
Rebased on master again. This time the failed build on Travis is definitely unrelated :slightly_smiling_face: 

Ping @nicolas-grekas, could you review again please?

fabpot commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -1093,18 +1094,21 @@ private function registerTranslatorConfiguration(array $config, ContainerBuilder
         $rootDir = $container->getParameter('kernel.root_dir');
         foreach ($container->getParameter('kernel.bundles_metadata') as $name => $bundle) {
-            if ($container->fileExists($dir = $bundle['path'].'/Resources/translations')) {
+            if (\file_exists($dir = $bundle['path'].'/Resources/translations')) {
```
Shouldn't we use `is_dir()` here instead of `file_exists()`?

rpkamp replied:
That would be better, yes. I'll change it!

fabpot commented on src/Symfony/Component/Translation/Translator.php:
```diff
@@ -395,7 +395,7 @@ private function getCatalogueCachePath($locale)
     }
 
-    private function doLoadCatalogue($locale): void
+    protected function doLoadCatalogue($locale): void
```
What about marking this method as `@internal` now that it is protected?

rpkamp replied:
Makes sense. I'll change that.

stof commented:
The list of files being loaded is still built by the container. So AFAICT, this change does not allow removing the resources at the container-level (we would still need to build the new list).
The only benefit of that change would be that the translator cache would be invalidated when adding new files too, solving #27600 (which is already a good move).
Some attempt was done in the past to include a hash of the list of resource files in the cache key, but this was reverted due to bad effects.

rpkamp commented:
> So AFAICT, this change does not allow removing the resources at the container-level (we would still need to build the new list).

This change is already removing the resources from the container, and everything works fine without the container knowing where translations files reside. Why would it need to know about those?

The only reason I can think of is if I add another translations path to the framework, but then the resource where I define that is not fresh anymore, so the container would be rebuilt, and the change would cascade to the catalogues.

rpkamp commented:
Comments processed @fabpot. Could you check again please?

curry684 commented:
Think it's right now but @fabpot should have a quick look still.

fabpot commented:
Thank you @rpkamp.

curry684 commented:
@fabpot label? ;)

fabpot commented:
@curry684 oops, fixed
